### DISCUSSION TITLE
Smoldot and runtime upgrades

### DISCUSSION THREAD

**Discussion by jsamol** [smol-dot/smoldot#1906] [All]
I'm working on integrating Smoldot into an Android application to communicate with a parachain. The nature of the application requires stable and reliable communication for extended periods without any user interaction or necessary app updates. This leads to the following questions, what is the expected behavior of the Smoldot client during a runtime upgrade? Can Smoldot handle runtime upgrades gracefully?

I am asking these questions because my team and I have already started running some tests. We have noticed that our application enters a semi-broken state where it appears to communicate with the network, fetch the state, and submit extrinsics, but not reliably. Specifically, we observed that some extrinsics sent periodically from our application are never broadcasted to the network, even though they were successfully validated by the client. Additionally, after the upgrade, I started noticing a significant increase in errors of the following type:

> Error while verifying justification: There exists a block in-between the latest finalized block and the block targeted by the justification that must first be finalized

At this moment, we do not have enough information to determine whether the issue lies with the client or on our side. We would appreciate any assistance in understanding what we should expect from the Smoldot client to further investigate the issue.

**Comment by tomaka** [comment]
Smoldot is supposed to handle runtime upgrades just fine.
Runtime upgrades are completely orthogonal to justifications or to justifications.
